internal/project/watch.go

```diff
@@ -60,3 +64,263 @@ func (w *watchedFiles[T]) update(ctx context.Context, newData T) (updated bool,
 	w.watcherID = watcherID
 	return true, nil
 }
+
+func createGlobMapper(host ProjectHost) func(data map[tspath.Path]string) []string {
+	rootDir := host.GetCurrentDirectory()
+	rootPath := tspath.ToPath(rootDir, "", host.FS().UseCaseSensitiveFileNames())
+	rootPathComponents := tspath.GetPathComponents(string(rootPath), "")
+	isRootWatchable := canWatchDirectoryOrFile(rootPathComponents)
+
+	return func(data map[tspath.Path]string) []string {
+		start := time.Now()
+
+		// dir -> recursive
+		globSet := make(map[string]bool)
+
+		for path, fileName := range data {
+			w := getDirectoryToWatchFailedLookupLocation(
+				fileName,
+				path,
+				rootDir,
+				rootPath,
+				rootPathComponents,
+				isRootWatchable,
+				rootDir,
+				true,
+			)
+			if w == nil {
+				continue
+			}
+			globSet[w.dir] = globSet[w.dir] || !w.nonRecursive
+		}
+
+		globs := make([]string, 0, len(globSet))
+		for dir, recursive := range globSet {
+			if recursive {
+				globs = append(globs, dir+"/"+recursiveFileGlobPattern)
+			} else {
+				globs = append(globs, dir+"/"+fileGlobPattern)
+			}
+		}
+		slices.Sort(globs)
+
+		timeTaken := time.Since(start)
+		host.Log(fmt.Sprintf("createGlobMapper took %s to create %d globs for %d failed lookups", timeTaken, len(globs), len(data)))
+		return globs
+	}
+}
+
+type directoryOfFailedLookupWatch struct {
+	dir            string
+	dirPath        tspath.Path
+	nonRecursive   bool
+	packageDir     *string
+	packageDirPath *tspath.Path
+}
+
+func getDirectoryToWatchFailedLookupLocation(
+	failedLookupLocation string,
+	failedLookupLocationPath tspath.Path,
+	rootDir string,
+	rootPath tspath.Path,
+	rootPathComponents []string,
+	isRootWatchable bool,
+	currentDirectory string,
+	preferNonRecursiveWatch bool,
+) *directoryOfFailedLookupWatch {
```

Comment:
Strada has baselines on what to watch : https://github.com/microsoft/TypeScript/blob/main/tests/baselines/reference/canWatch/getDirectoryToWatchFailedLookupLocationIndirDos.baseline.md like these. do we want these atleast till we have this logic and dont make this any faster.

Reply:
We can do that, though I'm wary of continuing to have these sorts of super large baselines.